core/src/main/kotlin/com/rudderstack/sdk/kotlin/core/SourceConfigManager.kt

```diff
@@ -68,25 +75,50 @@ class SourceConfigManager(
     private suspend fun downloadSourceConfig(): SourceConfig? {
         return withContext(analytics.networkDispatcher) {
             safelyExecute {
-                when (val sourceConfigResult = httpClientFactory.getData()) {
-                    is Result.Success -> {
-                        LenientJson.decodeFromString<SourceConfig>(sourceConfigResult.response).let { config ->
-                            LoggerAnalytics.info("SourceConfig is fetched successfully: $config")
-                            config
-                        }
-                    }
-
-                    is Result.Failure -> {
-                        LoggerAnalytics.error(
-                            "Failed to get sourceConfig due to ${sourceConfigResult.error}"
-                        )
-                        null
-                    }
+                getSourceConfigWithFallback {
+                    handleFailure(it)
                 }
             }
         }
     }
 
+    private suspend fun handleFailure(result: Result.Failure<NetworkErrorStatus>): SourceConfig? = when (result.error) {
+        NetworkErrorStatus.ERROR_400 -> {
+            analytics.shutdown()
+            null
+        }
+
+        NetworkErrorStatus.ERROR_401,
+        NetworkErrorStatus.ERROR_404,
+        NetworkErrorStatus.ERROR_413,
+        NetworkErrorStatus.ERROR_RETRY,
+        NetworkErrorStatus.ERROR_UNKNOWN,
+        NetworkErrorStatus.ERROR_NETWORK_UNAVAILABLE -> retryBlockWithBackoff {
+            getSourceConfigWithFallback { null }
+        }
```

Comment:
Let’s simplify this block to reduce lambda chaining and improve code readability.